### PR TITLE
Expose the record writer for validating arbitrary files

### PR DESCRIPTION
**Pull request by godlygeek**
Add a `RecordWriterTestHarness` class for testing and validating Memray record files by allowing direct writing of test records. This provides a way to create controlled test files with known contents for validating the file format and reader functionality. The implementation supports writing all record types including memory snapshots, allocations, frames, and thread information, making it possible to test edge cases and complex scenarios that would be difficult to trigger in normal operation.

### CODE REVIEW

**Review by pablogsal — APPROVED** [review]
Exquisite

File: src/memray/_memray/record_reader.cpp [hunk 1]
```diff
@@ -1100,10 +1100,11 @@ RecordReader::dumpAllRecordsFromAllAllocationsFile()
                             "<unknown allocator " + std::to_string((int)record.allocator) + ">";
                     allocator = unknownAllocator.c_str();
                 }
-                printf("address=%p size=%zd allocator=%s\n",
+                printf("address=%p size=%zd allocator=%s native_frame_id=%zd\n",
```
Comment (pablogsal) [inline src/memray/_memray/record_reader.cpp:1103]: Oh, good catch!
